report/sections/Chapter 07 - Queue System/Chapter 7.tex

```diff
@@ -1 +1,81 @@
+async fn worker_thread(
+    rx: Receiver<TestRunnerWork>,
+    job_results: Arc<Mutex<Box<HashMap<Uuid, Option<TestRunnerResult>>>>>,
+    limit : usize
+) {
+    let stream = tokio_stream::wrappers::ReceiverStream::new(rx);
+    
+    stream.for_each_concurrent(limit, |work| async {
+        debug!("Running worker thread on UUID {}...", work.id); |\label{line:concurrent-start}|
+
+        let res = schedule_test(work.submission).await;
+        let mut map = job_results.lock().unwrap();
+        map.insert(work.id, Some(res)); |\label{line:concurrent-end}|
+    }).await;
+}
+\end{lstlisting}
+
+We use the \texttt{futures} package in Rust to allocate some number of threads dictated by the variable \texttt{limit} of type \texttt{usize}.
+The receiver channel is converted to a \texttt{ReceiverStream} using the \texttt{tokio\_stream} utility pacckage.
+A \texttt{ReceiverStream} is simply a wrapper which implements the \texttt{Stream} trait from the \texttt{futures} package.
+This allows us to combine functionality from the two packages.
+In practice, implementations of the \texttt{Stream} trait simply provide a stream of values asynchronously --- essentially an asynchronous iterable.
+
+Using this, we can call \texttt{for\_each\_concurrent} from the \texttt{futures} package to run a number of tasks asynchronously but with the aforementioned limit. 
+The \texttt{futures} framework then manages the thread pool automatically, and we only need to specifiy the work that the threads need to do, which can be seen in lines \ref{line:concurrent-start} --- \ref{line:concurrent-end}.
+
+The \texttt{schedule\_test} function calls the test runner, which is responsible for running the GHC interpreter and returning the results of running the tests on the submitted code from the client, as described in chapter \ref{chap:TestRunner}.
+Once the result is computed, the thread acquires the mutex lock to the hashmap or blocks until it can acquire the lock.
+It then inserts the result into the hashmap by using the UUID as the key.
+Once the \texttt{map} variable goes out of scope, the mutex is unlocked so other threads may access the hashmap.
+
+In summary, a specified number of concurrent threads in a thread pool continuously receive code submissions from the \texttt{ReceiverStream} channel.
+Each submission is then processed by one of these threads, and the result of the test runner is packaged into the hashmap where the UUID serves as a key to look up the result.
+
+While this approach solved the initial problem of only being able to process one code submission at a time, the astute reader may have considered the fact that the hashmap continuously grows over time --- data is only ever added to the hashmap and never removed.
+Consequently, the memory use of the web server would eventually exceed its capacity, resulting in a crash.
```

Comment:
We call it web server, but it's the test runner to the readers. Otherwise they'll confuse the Next.js backend with the test runner.